roles/lib_openshift/src/class/oc_configmap.py

```diff
@@ -0,0 +1,183 @@
+                 name,
+                 from_file,
+                 from_literal,
+                 state,
+                 namespace,
+                 kubeconfig='/etc/origin/master/admin.kubeconfig',
+                 verbose=False):
+        ''' Constructor for OpenshiftOC '''
+        super(OCConfigMap, self).__init__(namespace, kubeconfig=kubeconfig, verbose=verbose)
+        self.name = name
+        self.state = state
+        self._configmap = None
+        self._inc_configmap = None
+        self.from_file = from_file if from_file is not None else {}
+        self.from_literal = from_literal if from_literal is not None else {}
+
+    @property
+    def configmap(self):
+        if self._configmap is None:
+            self._configmap = self.get()
+
+        return self._configmap
+
+    @configmap.setter
+    def configmap(self, inc_map):
+        self._configmap = inc_map
+
+    @property
+    def inc_configmap(self):
+        if self._inc_configmap is None:
+            results = self.create(dryrun=True, output=True)
+            self._inc_configmap = results['results']
+
+        return self._inc_configmap
+
+    @inc_configmap.setter
+    def inc_configmap(self, inc_map):
+        self._inc_configmap = inc_map
+
+    def from_file_to_params(self):
+        '''return from_files in a string ready for cli'''
+        return ["--from-file={}={}".format(key, value) for key, value in self.from_file.items()]
+
+    def from_literal_to_params(self):
+        '''return from_literal in a string ready for cli'''
+        return ["--from-literal={}={}".format(key, value) for key, value in self.from_literal.items()]
+
+    def get(self):
+        '''return a configmap by name '''
+        results = self._get('configmap', self.name)
+        if results['returncode'] == 0 and results['results'][0]:
+            self.configmap = results['results'][0]
+
+        if results['returncode'] != 0 and '"{}" not found'.format(self.name) in results['stderr']:
+            results['returncode'] = 0
+
+        return results
+
+    def delete(self):
+        '''delete a configmap by name'''
+        return self._delete('configmap', self.name)
+
+    def create(self, dryrun=False, output=False):
+        '''Create a configmap
+
+           :dryrun: Product what you would have done. default: False
+           :output: Whether to parse output. default: False
+        '''
+
+        cmd = ['create', 'configmap', self.name]
+        if self.from_literal is not None:
+            cmd.extend(self.from_literal_to_params())
+
+        if self.from_file is not None:
+            cmd.extend(self.from_file_to_params())
+
+        if dryrun:
+            cmd.extend(['--dry-run', '-ojson'])
+
+        results = self.openshift_cmd(cmd, output=output)
+
+        return results
+
+    def update(self):
+        '''run update configmap '''
+        return self._replace_content('configmap', self.name, self.inc_configmap)
+
+    def needs_update(self):
```

Comment:
The biggest problem with `oc replace` (doing an `UPDATE`) is that the resource version is not required, and when it is set to `0` or omitted, as is often with serialized config on disk you want to use to drive state, the operation clobbers data on the server and can be dangerous.

In cases where the installer or the user or whatever is driving the client knows for certain that it controls the entire document and can determine the state of the entire document at once, there is no problem. However, once the user is attempting to convey the intent of "I want to make this edit" it is not going to work as you would expect. Consider this case:

 - user `kenny` reads a resource off of the server
 - user `kenny` makes some local edits to the resource on his local machine
 - user `steve` updates the resource on the server
 - user `kenny` uses `oc replace` with his edited version

If the intent here was that `kenny` would make an edit to the object, this is the wrong flow. In this case, even if `steve` made edits to a completely unrelated portion of the object, `oc replace` will clobber the entirety of the object with what user `kenny` posted. 

Your points re: the "did I update" semantic are valid and should be brought up in Kubernetes CLI SIG. Perhaps also the question should be fielded on how the upstream Ansible modules are working with this. But in either case when a user wants to update a portion of an object, and does not know for certain that they control the entirety of the object, the `oc replace` is a dangerous and incorrect tool to use. Providing an explicit resource version can help detect places where unknown clobber would occur but in general you'd still be re-pulling, re-editing and re-pushing (and may race with someone else again) but if you just submitted a patch you'd be fine.